flytekit/core/rust/src/lib.rs

```diff
@@ -0,0 +1,208 @@
+            .build(),
+    };
+
+    Client::from_conf(s3_conf)
+}
+
+impl S3FileSystem {
+    fn get_client(&self) -> &aws_sdk_s3::Client {
+        &self.s3_client
+    }
+}
+
+#[pymethods]
+impl S3FileSystem {
+    #[new]
+    pub fn new(endpoint: String) -> S3FileSystem {
+        let c = build_client(&endpoint);
+        S3FileSystem {
+            endpoint: endpoint,
+            s3_client: c,
+        }
+    }
+
+    pub fn put_file(&self, lpath: String, bucket: String, key: String) -> PyResult<()> {
+        let client = self.get_client();
+        let rt = tokio::runtime::Runtime::new().unwrap();
+        rt.block_on(async {
+            let file_size = std::fs::metadata(&lpath).unwrap().len();
+            if file_size < cmp::min(5 * 2u64.pow(30), 2 * CHUNK_SIZE) {
+                let body = ByteStream::from_path(lpath).await;
+                client
+                    .put_object()
+                    .bucket(bucket)
+                    .key(key)
+                    .body(body.unwrap())
+                    .send()
+                    .await
+                    .unwrap();
+            } else {
+                let mut chunk_count = (file_size / CHUNK_SIZE) + 1;
+                let mut size_of_last_chunk = file_size % CHUNK_SIZE;
+                if size_of_last_chunk == 0 {
+                    size_of_last_chunk = CHUNK_SIZE;
+                    chunk_count -= 1;
+                }
+
+                let multipart_upload_res: CreateMultipartUploadOutput = client
+                    .create_multipart_upload()
+                    .bucket(&bucket)
+                    .key(&key)
+                    .send()
+                    .await
+                    .unwrap();
+                let upload_id = multipart_upload_res.upload_id.unwrap();
+
+                // Create futures for each chunk upload
+                let upload_futures: Vec<_> = (0..chunk_count).map(|chunk_index| {
+                    let client = &client;
+                    let lpath = &lpath;
+                    let bucket = &bucket;
+                    let key = &key;
+                    let upload_id = &upload_id;
+
+                    async move {
+                        let this_chunk = if chunk_count - 1 == chunk_index {
+                            size_of_last_chunk
+                        } else {
+                            CHUNK_SIZE
+                        };
+                        let stream = ByteStream::read_from()
+                            .path(lpath)
+                            .offset(chunk_index * CHUNK_SIZE)
+                            .length(Length::Exact(this_chunk))
+                            .build()
+                            .await
+                            .unwrap();
+
+                        // Chunk index needs to start at 0, but part numbers start at 1.
+                        let part_number = (chunk_index as i32) + 1;
+                        let upload_part_res = client.upload_part()
```

Comment:
how are errors handled?
If you see S3FS handles errors
https://github.com/fsspec/s3fs/blob/a14c31a382c50996aff37107245768a13efef0cb/s3fs/core.py#L108-L128

can you test with no s3 connection or kill the connection in between, how does that work
or in case we get a slowdown or some other retryable error

Reply:
Based on what I know, the error in rust will be converted to Python error, but I will spent some time looking into it.

Of course I can test that. I will post the result later